### PR TITLE
Add 100-500ms of per-peer inventory delay

### PR DESCRIPTION
**Pull request by jrick**
Similar to a recent dcrd change, we want published inventory by SPV wallet peers be submitted with a small random amount of delay.  In this change we use the same logic that dcrd now uses to pick, for each connected peer, a random delay between 100-500ms before the inventory is sent.

### CODE REVIEW

File: p2p/peering.go [hunk 1]
```diff
@@ -497,6 +511,51 @@ func (mw *msgWriter) write(ctx context.Context, msg wire.Message, pver uint32) e
 	}
 }
 
+func (rp *RemotePeer) sendInventory(ctx context.Context) error {
+	timeout := func() time.Duration {
+		const d = maxInvTrickleTimeout - minInvTrickleTimeout
+		return minInvTrickleTimeout + rand.Duration(d)
+	}
+	timer := time.NewTimer(timeout())
```
Comment (matthawkins90) [inline p2p/peering.go:519]: What's the point of setting the variable `timer` here? It doesn't look like it's used, and you're already using `timer.Reset(timeout())` below on line 526.

**Review by matthawkins90 — COMMENTED** [review]
Something I noticed in https://github.com/decred/dcrd/pull/3363 was that the address list was shuffled.

Should you also shuffle the address list here?

I know you said:

> Currently, with every peer using a 500ms ticker, if no changes occur to the graph of connected nodes, messages will always propagate the nodes in the same order. However, if all nodes are randomizing their inventory batching delays, the next peer who will relay a message first during the next hop will always be different.

So it's possible that shuffling the address list is redundant, but I figured I'd ask if it was worth doing.

File: p2p/peering.go (hunk 1 above)
Comment (matthawkins90) [reply p2p/peering.go:519]: Oh I'm sorry, I spoke too soon. You have to wait for `timer.C`, and then you reset it with a new randomized `timeout()`. I misread it as doubling-up the time-duration.

**Comment by jrick** [comment]
> Something I noticed in https://github.com/decred/dcrd/pull/3363 was that the address list was shuffled.
>
> Should you also shuffle the address list here?

dcrwallet does not respond to `getaddr`.  It only requests them.

**Comment by jrick** [comment]
or did you mean this, shuffling the inventory before we publish?

```
diff /home/jrick/src/dcrwallet
commit - b226415203f66ccaaad9d5e70bdb5adfa2e2e565
path + /home/jrick/src/dcrwallet
blob - 6e4effdee7b8a6306d5b60586d3a498b605aa6a6
file + p2p/peering.go
--- p2p/peering.go
+++ p2p/peering.go
@@ -542,6 +542,7 @@ func (rp *RemotePeer) sendWaitingInventory(ctx context
 	}
 
 	s := rp.waitingInvs
+	rand.ShuffleSlice(s)
 	for len(s) != 0 {
 		l := len(s)
 		if l > wire.MaxInvPerMsg {
```

**Comment by matthawkins90** [comment]
Yes, sorry. Shuffling the inventory. Just wondering if it was worth doing.

**Comment by jrick** [comment]
it's a good idea. we shouldn't need to depend on the order being dependency order, as both mempool and mixpool handle orphans.

File: p2p/peering.go [hunk 2]
```diff
@@ -2003,20 +2071,20 @@ func (rp *RemotePeer) HeadersAsync(ctx context.Context, blockLocators []*chainha
 // hashes of txs.
 func (rp *RemotePeer) PublishTransactions(ctx context.Context, txs ...*wire.MsgTx) error {
 	const opf = "remotepeer(%v).PublishTransactions"
-	inv := wire.NewMsgInvSizeHint(uint(len(txs)))
+
+	// This would be better written to immediately send any vote
+	// transactions and sidestepping the inv batching, however SPV wallets
+	// cannot vote so the liklihood of them needing to ever send these
```
Comment (matthawkins90) [inline p2p/peering.go:2077]: sp: Likelihood

File: p2p/peering.go [hunk 3]
```diff
@@ -2003,20 +2071,20 @@ func (rp *RemotePeer) HeadersAsync(ctx context.Context, blockLocators []*chainha
 // hashes of txs.
 func (rp *RemotePeer) PublishTransactions(ctx context.Context, txs ...*wire.MsgTx) error {
 	const opf = "remotepeer(%v).PublishTransactions"
```
Comment (matthawkins90) [inline p2p/peering.go:2073]: now that `errors.Opf()` is removed from this function, this const seems unused?

**Review by matthawkins90 — COMMENTED** [review]
I like the changes to `PublishTransactions` and `PublishMixMessages`. A lot easier to read.

**Review by matthawkins90 — APPROVED** [review]

**Comment by davecgh** [comment]
> it's a good idea. we shouldn't need to depend on the order being dependency order, as both mempool and mixpool handle orphans.

I actually disagree with this.  In fact, it's a bad idea, imo.  It's true that mempool handles orphans, but only a pretty limited number of them on purpose to prevent malicious actors being able to fill up memory without actually spending any real coins.  It assumes that they are going to mostly be advertised in order.  Changing that would make it so malicious actors can hinder propagation by intentionally sending a bunch of junk to flush orphan buffers.

File: p2p/peering.go [hunk 4]
```diff
@@ -497,6 +511,52 @@ func (mw *msgWriter) write(ctx context.Context, msg wire.Message, pver uint32) e
...
+	defer rp.waitingInvsMu.Unlock()
+
+	sendInvMsg := func(invList []*wire.InvVect) error {
+		msg := new(wire.MsgInv)
+		msg.InvList = invList
+		return rp.sendMessageAck(ctx, msg)
+	}
+
+	s := rp.waitingInvs
+	rand.ShuffleSlice(s)
```
Comment (davecgh) [inline p2p/peering.go:545]: As mentioned, this opens up easy ways for malicious peers to hinder propagation.  Please don't intentionally send invs out of dependency order.

**Review by davecgh — CHANGES_REQUESTED** [review]

File: p2p/peering.go (hunk 4 above)
Comment (matthawkins90) [reply p2p/peering.go]: Sorry bout that

**Review by davecgh — APPROVED** [review]

**Pull request merged**